src/osd/OSD.cc

```diff
@@ -4679,6 +4679,30 @@ struct C_OSD_GetVersion : public Context {
   }
 };
 
+bool OSD::ms_handle_refused(Connection *con)
+{
+  OSD::Session *session = (OSD::Session *)con->get_priv();
+  dout(1) << "ms_handle_refused con " << con << " session " << session << dendl;
+  if (!session)
+    return false;
+  int id = con->get_peer_id();
+  int type = con->get_peer_type();
+  // handle only OSD failures here
+  if (monc && (type == CEPH_ENTITY_TYPE_OSD)) {
+    OSDMapRef osdmap = get_osdmap();
+    if (osdmap && osdmap->is_up(id)) {
+      // I'm cheating mon heartbeat grace logic, because we know it's not going
+      // to respawn alone. +1 so we won't hit any boundary case.
+      monc->send_mon_message(new MOSDFailure(monc->get_fsid(),
+                                                osdmap->get_inst(id),
+                                                cct->_conf->osd_heartbeat_grace + 1,
+                                                osdmap->get_epoch()
+                                                ));
```

Comment:
On Fri, Apr 22, 2016 at 09:28:13AM -0700, Sage Weil wrote:
> Let's update teh message properly.. something like
> 
> [..]
>   public:
> +  enum {
> +    FLAG_FAILED = 1,     // if set, failure; if not, recovery
> +    FLAG_IMMEDIATE = 2,  // known failure, not a timeout
> +  };
> +  
> [..] 
> then we can use this for other stuff too, like a watchdog.

And introduce other daemon failure modes, too.

-- 
Piotr Dałek